From 6bd4384a8bdcad5af1a7ba323edb7da9251eca79 Mon Sep 17 00:00:00 2001
From: John Johansen <john.johansen@canonical.com>
Date: Fri, 18 Mar 2016 06:06:06 -0700
Subject: [PATCH 1625/1767] Revert "UBUNTU: SAUCE: apparmor: Fix: insert race
 between label_update and label_merge"

BugLink: http://bugs.launchpad.net/bugs/1379535

This reverts commit c8cf841fc8307461847036d2be2ad27791be485b.

Signed-off-by: Tim Gardner <tim.gardner@canonical.com>
---
 security/apparmor/label.c | 11 ++---------
 1 file changed, 2 insertions(+), 9 deletions(-)

diff --git a/security/apparmor/label.c b/security/apparmor/label.c
index d08f711..711c632 100644
--- a/security/apparmor/label.c
+++ b/security/apparmor/label.c
@@ -1782,22 +1782,15 @@ static struct aa_label *__label_update(struct aa_label *label)
 	if (invcount) {
 		l->size -= aa_sort_and_merge_profiles(l->size, &l->ent[0]);
 		if (labels_set(label) == labels_set(l)) {
-			goto insert;
+			AA_BUG(__aa_label_remove_and_insert(labels_set(label), label, l) != l);
 		} else {
 			aa_label_remove(labels_set(label), label);
 			goto other_ls_insert;
 		}
 	} else {
 		AA_BUG(labels_ns(label) != labels_ns(l));
-		goto insert;
+		AA_BUG(__aa_label_remove_and_insert(labels_set(label), label, l) != l);
 	}
-insert:
-	tmp = __aa_label_remove_and_insert(labels_set(label), label, l);
-	if (tmp != l) {
-		aa_get_label(tmp);
-		aa_label_free(l);
-	 }
-
 	write_unlock_irqrestore(&ls->lock, flags);
 
 	return l;
-- 
1.9.3

